# Proyecto 1 – Predecir la calidad del Vino

### Integrantes
- Camilo Andres Galeano Trujillo

### Problema
Por medio de una base de datos con información referente al vino rojo de la compañia "Vinho Verde" se quiere saber la calidad de estos, siendo una variable categórica ordinal con puntajes del 1 al 10, encontramos las siguiente variables para dar paso a esa categorización:

- fixed acidity
- volatile acidity
- citric acid
- residual sugar
- chlorides
- free sulfur dioxide
- total sulfur dioxide
- density
- pH
- sulphates
- alcohol

### Enlace GitHub

<https://github.com/cgatrujillo/ml-ean>

In [10]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.metrics import confusion_matrix, classification_report, mean_absolute_error, mean_squared_error

In [3]:
df = pd.read_csv("winequality-red.csv", sep=";")
df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


Los datos consisten principalmente en las variables químicas que componen cada uno de los 1599 vinos rojos registrados, de las doce variables presentadas 11 son de tipo númerico flotante y la última que nos indica la calidad es categórica ordinal.

In [5]:
X = df.drop(columns="quality")
y = df["quality"]

In [26]:
model_rl = Pipeline(
    steps= [("scaler", StandardScaler()), ("reglin", LinearRegression())] 
)

In [27]:
model_rl.fit(X, y)
print("Modelo Regresion Lineal: %3f" % model_rl.score(X, y))

Modelo Regresion Lineal: 0.360552


In [28]:
y_predict = model_rl.score(X, y)
print("Coefficient of determination in the training set: ", y_predict)
print("MAE: ", mean_absolute_error(y, model_rl.predict(X)))
print("MSE: ", mean_squared_error(y, model_rl.predict(X)))
print("RMSE: ", np.sqrt(mean_squared_error(y, model_rl.predict(X))))

Coefficient of determination in the training set:  0.3605517030386881
MAE:  0.5004899635644883
MSE:  0.41676716722140805
RMSE:  0.6455750670692046


El modelo de regresión lineal sencillo, no está consiguiendo interpretar de manera adecuada la calidad de los vinos y las predicciones realizadas presentan una gran desviación.

In [20]:
df["categorical_quality"] = df['quality'].map(lambda x: 'Bueno' if x >= 6 else 'Regular')

In [22]:
Xtrain, Xtest, ytrain, ytest = train_test_split(df.drop(columns=["quality", "categorical_quality"]), df["categorical_quality"], test_size =0.2, random_state=12)

In [24]:
Xtrain.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count    1279.000000       1279.000000  1279.000000     1279.000000   
mean        8.312119          0.528292     0.270727        2.532604   
std         1.727279          0.179117     0.194901        1.366531   
min         4.600000          0.120000     0.000000        0.900000   
25%         7.100000          0.390000     0.090000        1.900000   
50%         7.900000          0.520000     0.260000        2.200000   
75%         9.200000          0.640000     0.420000        2.600000   
max        15.900000          1.330000     1.000000       15.500000   

         chlorides  free sulfur dioxide  total sulfur dioxide      density  \
count  1279.000000          1279.000000           1279.000000  1279.000000   
mean      0.086908            15.662627             46.562940     0.996742   
std       0.046470            10.088504             33.197896     0.001896   
min       0.012000             1.000000              6.000000     0.990070   
25%       0.070000             7.000000             22.000000     0.995600   
50%       0.079000            14.000000             38.000000     0.996760   
75%       0.090000            21.000000             63.000000     0.997860   
max       0.611000            68.000000            289.000000     1.003690   

                pH    sulphates      alcohol  
count  1279.000000  1279.000000  1279.000000  
mean      3.312361     0.658288    10.431744  
std       0.154791     0.169065     1.083963  
min       2.740000     0.330000     8.400000  
25%       3.210000     0.550000     9.500000  
50%       3.310000     0.620000    10.200000  
75%       3.400000     0.730000    11.100000  
max       4.010000     2.000000    14.900000

In [25]:
Xtest.describe()

fixed acidity  volatile acidity  citric acid  residual sugar  \
count     320.000000        320.000000   320.000000      320.000000   
mean        8.349687          0.525938     0.271969        2.563594   
std         1.797717          0.179098     0.194705        1.573555   
min         5.000000          0.160000     0.000000        1.300000   
25%         7.100000          0.400000     0.100000        1.900000   
50%         7.950000          0.520000     0.250000        2.100000   
75%         9.300000          0.636250     0.430000        2.600000   
max        15.600000          1.580000     0.760000       15.400000   

        chlorides  free sulfur dioxide  total sulfur dioxide     density  \
count  320.000000           320.000000            320.000000  320.000000   
mean     0.089700            16.723437             46.087500    0.996767   
std      0.049386            11.808925             31.704827    0.001855   
min      0.012000             1.000000              7.000000    0.990640   
25%      0.071000             8.000000             22.000000    0.995615   
50%      0.079000            13.000000             38.500000    0.996665   
75%      0.091000            24.000000             60.000000    0.997755   
max      0.415000            72.000000            160.000000    1.003690   

               pH   sulphates     alcohol  
count  320.000000  320.000000  320.000000  
mean     3.306125    0.657594   10.387969  
std      0.152897    0.171527    0.990007  
min      2.880000    0.390000    8.500000  
25%      3.200000    0.550000    9.500000  
50%      3.310000    0.620000   10.100000  
75%      3.410000    0.732500   11.000000  
max      3.900000    1.610000   14.000000

In [29]:
model_lr = Pipeline(
    steps= [("scaler", StandardScaler()), ("logistic", LogisticRegression())] 
)

In [31]:
ypred = model_lr.predict(Xtest)
confusion_matrix(ytest, ypred)

array([[134,  40],
       [ 43, 103]], dtype=int64)

In [33]:
print(classification_report(ytest, ypred))

              precision    recall  f1-score   support

       Bueno       0.76      0.77      0.76       174
     Regular       0.72      0.71      0.71       146

    accuracy                           0.74       320
   macro avg       0.74      0.74      0.74       320
weighted avg       0.74      0.74      0.74       320



Comparando ambos modelos, podemos deducir que el modelo con mejor funcionamiento es el de regresión logística, viendo las métricas como el coeficiente de determinación del modelo lineal, que es bastante bajo, un valor de 0.36 y el valor de accuracy del modelo logístico en un 0.74, lo cual es más aceptable, a su vez es un modelo se podría mejorar aplicando regularización, validación cruzada y la busqueda de sus mejores hiperparámetros. También podemos tener en cuenta que el modelo logístico presenta una mejor posibilidad de clasificación adecuada en los vinos buenos que los regulares y que en el modelo lineal tuvimos en cuenta 11 posibles clasificaciones de calidad en vez de un resultado binario como el logístico.